File: Project.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr 21 00:10:03 2019

@author: vinayak
"""

#Load libraries

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency
import seaborn as sns
from random import randrange, uniform
from sklearn.cross_validation import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

#Set working directory
os.chdir("C:\Users\vinayak\Desktop\Bike Renting Project")

#Load data
dataset = pd.read_csv("day.csv")


##################################Missing value analysis################################################
missing_val = pd.DataFrame(dataset.isnull().sum())
#There are no missing values in the dataset so no need to perform missing value analysis.

##################################Feature Selection################################################
#Remove the variable that are not useful for the analysis
#1. Instant there is no need to add it as it only explains the row number
#2. Dteday there is no need to add date as year and month are already present in 2 different columns and date 
#   does not have large impact on the result
#3. Casual and Registered: Ideally we should not use these 2 variable because cnt = casual and Registered
#   and we need to calculate the bike count on the basis of environmental and seasonal settings.

dataset = dataset.drop(["instant","dteday","casual","registered"], axis=1) 

#save numeric names
cnames =  ["temp","atemp","hum","windspeed"]

df_corr = dataset.loc[:,cnames]
#Set the width and hieght of the plot
f, ax = plt.subplots(figsize=(7, 5))
#Generate correlation matrix
corr = df_corr.corr()
#Plot using seaborn library
sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap=sns.diverging_palette(220, 10, as_cmap=True),
            square=True, ax=ax)

#As we can see in the plot that that temp and atemp are very high +vely correlated then we can drop any 1 
#So here I am dropping atemp.
dataset = dataset.drop(['atemp'],axis = 1)

##################################Outlier Analysis################################################

cnames =  ["temp","hum","windspeed"]
plt.boxplot(dataset['temp'])
plt.boxplot(dataset['hum'])
plt.boxplot(dataset['windspeed'])
#Detect and replace with NA
#Extract quartiles
for i in cnames:
    q75, q25 = np.percentile(dataset.loc[:,i], [75 ,25])
    #Calculate IQR
    iqr = q75 - q25
    #Calculate inner and outer fence
    minimum = q25 - (iqr*1.5)
    maximum = q75 + (iqr*1.5)
    #Replace with NA
    dataset.loc[dataset.loc[:,i] < minimum,i] = np.nan
    dataset.loc[dataset.loc[:,i] > maximum,i] = np.nan
     
missing_val = pd.DataFrame(dataset.isnull().sum())

dataset['hum'] = dataset['hum'].fillna(dataset['hum'].mean())
dataset['windspeed'] = dataset['windspeed'].fillna(dataset['windspeed'].mean())


##################################Feature scaling################################################
# In the dataset there are only 3 numeric variables temp, hum, windspeed and the value of these variable 
# are in the range of 0 and 1, so no need to apply feature scaling.

###################################Model Development#######################################

#Divide data into train and test set
train, test = train_test_split(dataset, test_size=0.2)

#Calculate MAPE
def MAPE(y_true, y_pred): 
    mape = np.mean(np.abs((y_true - y_pred) / y_true))*100
    return mape

#Decision tree for regression
regressor = DecisionTreeRegressor(max_depth =6, random_state = 0)
regressor.fit(train.iloc[:,0:10], train.iloc[:,10])             #Fit the model on training data
y_pred = regressor.predict(test.iloc[:,0:10])                   #predicting the test set results
MAPE(test.iloc[:,10], y_pred)                                   #Compare with actual

#Error Rate: 15.91
#Accuracy: 84.09

#Linear Regression
regressor = LinearRegression()     
regressor.fit(train.iloc[:,0:10], train.iloc[:,10])             #Fit the model on training data
y_pred = regressor.predict(test.iloc[:,0:10])                   #predicting the test set results
MAPE(test.iloc[:,10], y_pred)                                   #Compare with actual

#Error Rate: 19.16
#Accuracy: 80.84

#Random Forest
regressor = RandomForestRegressor(n_estimators = 300, random_state = 0)
regressor.fit(train.iloc[:,0:10], train.iloc[:,10])             #Fit the model on training data
y_pred = regressor.predict(test.iloc[:,0:10])                   #predicting the test set results
MAPE(test.iloc[:,10], y_pred)                                   #Compare with actual

#Error Rate: 11.72
#Accuracy: 88.28

#NOTE: If you run the Model again and again then you might see some slight difference in the accuracy because we are 
# creating the test and training set randomly, because of this, accuracy difference would be there after every run.

File: Project.R
#Clear the environment
rm(list=ls())

#set the working directory
setwd(dir = "C:/Users/vinayak/Desktop/Bike Renting Project")

#Load Libraries
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", 'sampling', 'DataCombine', 'inTrees',"createDataPartition","usdm","randomForest","e1071")

install.packages(x)
lapply(x, require, character.only = TRUE)
rm(x)

#Load the dataset
dataset = read.csv("day.csv")

#Check the structure of the dataset
str(dataset)

##################################Missing value analysis################################################
sum(is.na(dataset))
#There are no missing values in the dataset so no need to perform missing value analysis.

#Univariate Analysis and Variable Consolidation
dataset$season = as.factor(dataset$season)
dataset$yr = as.factor(dataset$yr)
dataset$mnth = as.factor(dataset$mnth)
dataset$holiday = as.factor(dataset$holiday)
dataset$weekday = as.factor(dataset$weekday)
dataset$workingday = as.factor(dataset$workingday)
dataset$weathersit = as.factor(dataset$weathersit)

#Remove the variable that are not useful for the analysis
#1. Instant there is no need to add it as it only explains the row number
#2. Dteday there is no need to add date as year and month are already present in 2 different columns and date 
#   does not have large impact on the result
#3. Casual and Registered: Ideally we should not use these 2 variable because cnt = casual and Registered
#   and we need to calculate the bike count on the basis of environmental and seasonal settings.

c = c(-1,-2,-14,-15)
dataset = dataset[c]

##################################Outlier Analysis################################################

numeric_index = sapply(dataset,is.numeric) #selecting only numeric
numeric_data = dataset[,numeric_index]

factor_index = sapply(dataset,is.factor)  #selecting only factor
factor_data = dataset[,factor_index]


cnames = colnames(numeric_data)
# 
for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "cnt"), data = subset(dataset))+
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="cnt")+
           ggtitle(paste("Box plot of cnt for",cnames[i])))
}
#
## Plotting plots together
gridExtra::grid.arrange(gn1,gn2,gn3,ncol=3)
gridExtra::grid.arrange(gn4,ncol=3)

for(i in cnames){
  val = dataset[,i][dataset[,i] %in% boxplot.stats(dataset[,i])$out]
  #print(length(val))
  dataset[,i][dataset[,i] %in% val] = NA
}
sum(is.na(dataset))

#Impute NA using KNN impute
dataset = knnImputation(dataset, k = 3)

##################################Feature Selection################################################
## Correlation Plot 
corrgram(dataset[,numeric_index], order = F,
         upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")


#As we can see in the plot that that temp and atemp are very high +vely correlated then we can drop any 1 
#So here I am dropping atemp.
dataset = dataset[,-9]

##################################Feature scaling################################################
# In the dataset there are only 3 numeric variables temp, hum, windspeed and the value of these variable 
# are in the range of 0 and 1, so no need to apply feature scaling.

###################################Model Development#######################################
#Clean the environment
rmExcept("dataset")


#Divide data into train and test set
set.seed(1234)
train_index = sample(1:nrow(dataset), 0.8 * nrow(dataset))
train = dataset[train_index,]
test = dataset[-train_index,]

#MAPE
#calculate MAPE
MAPE = function(y, yhat){
  mean(abs((y - yhat)/y))*100
}


#rpart for regression
fit = rpart(cnt ~ ., data = train, method = "anova")
predictions_DT = predict(fit, test[,-11])     #Predict for new test cases
MAPE(test[,11], predictions_DT)

#Error Rate: 21.98337
#Accuracy: 78.01663

#run regression model
lm_model = lm(cnt ~., data = train)
predictions_LR = predict(lm_model, test[,-11])    #Predict
MAPE(test[,11], predictions_LR)                   #Calculate MAPE

#Error Rate: 20.47
#acuracy: 79.53%

# Fitting Random Forest Regression to the dataset
set.seed(1234)
regressor = randomForest(x = train[-11], y = train$cnt,ntree = 500)
predictions_regressor = predict(regressor, test[,-11])
MAPE(test[,11], predictions_regressor)

#Error Rate: 19.58
#acuracy: 80.42%

# Fitting SVR to the dataset
regressor = svm(formula = cnt ~ .,data = train,type = 'eps-regression',kernel = 'radial')
predictions_regressor = predict(regressor, test[,-11])
MAPE(test[,11], predictions_regressor)

#Error Rate: 16.44
#acuracy: 83.56%

#Pass sample input into the model
sample = data.frame(season=1,	yr=0,	mnth=1,	holiday=0,	weekday=1,	workingday=1,	weathersit=1,	
                    temp=0.160833,	hum=0.492917, windspeed=0.223267, cnt=1321)
newtest = rbind(test, sample)
predict(regressor, newtest[148,-11])


#You might see some slight differences in the accuracy as i applied outlier anlysis quite a few times
#to remove outliers.